interfaces/apparmor/backend_test.go

```diff
@@ -492,7 +495,362 @@ func (s *backendSuite) TestSetupHostSnapConfineApparmorForReexecWritesNew(c *C)
+	data, err := ioutil.ReadFile(filepath.Join(dirs.SnapConfineAppArmorDir, files[0].Name()))
+	c.Assert(err, IsNil)
+	c.Assert(string(data), testutil.Contains, "network inet,")
+	c.Assert(string(data), testutil.Contains, "network inet6,")
+
+	// The distribution policy was not reloaded because snap-confine executes
+	// from core snap. This is handled separately by per-profile Setup.
+	c.Assert(cmd.Calls(), HasLen, 0)
+}
+
+// Test behavior when issHomeUsingNFS fails.
+func (s *backendSuite) TestSetupSnapConfineGeneratedPolicyError1(c *C) {
+	// Make corrupt mountinfo.
+	restore := apparmor.MockMountInfo("corrupt")
+	defer restore()
+
+	// Intercept interaction with apparmor_parser
+	cmd := testutil.MockCommand(c, "apparmor_parser", "")
+	defer cmd.Restore()
+
+	// Intercept the /proc/self/exe symlink and point it to the snapd from the
+	// mounted core snap. This indicates that snapd has re-executed and
+	// should not reload snap-confine policy.
+	fakeExe := filepath.Join(s.RootDir, "corrupt-proc-self-exe")
+	restore = apparmor.MockProcSelfExe(fakeExe)
+	defer restore()
+
+	// Setup generated policy for snap-confine.
+	err := apparmor.SetupSnapConfineGeneratedPolicy()
+	// NOTE: Errors in determining NFS are non-fatal to prevent snapd from
+	// failing to operate. A warning message is logged but system operates as
+	// if NFS was not active.
+	c.Assert(err, IsNil)
+
+	// While other stuff failed we created the policy directory and didn't
+	// write any files to it.
+	files, err := ioutil.ReadDir(dirs.SnapConfineAppArmorDir)
+	c.Assert(err, IsNil)
+	c.Assert(files, HasLen, 0)
+
+	// We didn't reload the policy.
+	c.Assert(cmd.Calls(), HasLen, 0)
+}
+
+// Test behavior when os.Readlink "/proc/self/exe" fails.
+func (s *backendSuite) TestSetupSnapConfineGeneratedPolicyError2(c *C) {
+	// Make it appear as if NFS was mounted under /home.
+	restore := apparmor.MockMountInfo("680 27 0:59 / /home/zyga/nfs rw,relatime shared:478 - nfs4 localhost:/srv/nfs rw,vers=4.2,rsize=524288,wsize=524288,namlen=255,hard,proto=tcp,port=0,timeo=600,retrans=2,sec=sys,clientaddr=127.0.0.1,local_lock=none,addr=127.0.0.1")
+	defer restore()
+	restore = apparmor.MockEtcFstab("")
+	defer restore()
+
+	// Intercept interaction with apparmor_parser
+	cmd := testutil.MockCommand(c, "apparmor_parser", "")
+	defer cmd.Restore()
+
+	// Intercept the /proc/self/exe symlink and point it to the snapd from the
+	// mounted core snap. This indicates that snapd has re-executed and
+	// should not reload snap-confine policy.
+	fakeExe := filepath.Join(s.RootDir, "corrupt-proc-self-exe")
+	restore = apparmor.MockProcSelfExe(fakeExe)
+	defer restore()
+
+	// Setup generated policy for snap-confine.
+	err := apparmor.SetupSnapConfineGeneratedPolicy()
+	c.Assert(err, ErrorMatches, "cannot read /proc/self/exe,.*")
+
+	// We created the policy file.
+	files, err := ioutil.ReadDir(dirs.SnapConfineAppArmorDir)
+	c.Assert(err, IsNil)
+	c.Assert(files, HasLen, 1)
```

Comment:
If we can't read /proc/self/exe, why would we create the policy file? We want to create the directory but I see no reason why we would fallback to creating the file, even if we don't reload, since something outside of snapd might reload the policy and then have the NFS policy included.

Reply:
Because the policy is equally applicable to the re-exec case. We generate the local policy file but don't reload the distro profile. When Setup is called it will choose (or not) to reload the core profile. If it chooses to reload it at that time the generated include file will be present and we will benefit from it.

In all cases I think that if we cannot examine /proc/self/exe we will have other trouble elsewhere so I think this is good as is.

Reply:
I understand with the way you wrote setupSnapConfineGeneratedPolicyImpl() why you wrote the test this way, but my point is to fail closed rather than open (and reorder the read on procSelfExe so we don't generate the policy. The idea is if there is a crazy kernel bug on reading /proc/self/exe, we fail early and loud rather than try to plod along.

I also understand your point that being able to read /proc/self/exe or not doesn't really have anything to do with isHomeUsingNFS(), which is why you wrote out the policy. I'll leave the decision up to you (ie, not a blocker).

Reply:
I'm changing this now.